### ISSUE TITLE
DLL calling is broken (wrong string and real values)

### ISSUE THREAD

**Issue by time-killer-games** [enigma-dev/enigma-dev#1511]
So this used to work for me, but in my most recent test all values whether returned or passed as a parameter to a DLL, it is pretty random whether the numbers or strings show with the correct values or not.

[DialogModule.zip](https://github.com/enigma-dev/enigma-dev/files/4253011/DialogModule.zip)

Run the test project attached above to see what I mean. This same DLL works just fine in GameMaker 8.1, GameMaker Studio 1.4, GameMaker Studio 2.2, Unity/C#, VB, VBScript, and JScript. Thus, this is an obvious bug with ENIGMA, and not the DLL itself. I can't reproduce this issue in any other engine or language. :-(

**Comment by RobertBColton** [comment]
Seems to be a duplicate of #1461 and the regression was probably introduced by #1445.

>Moved the declaration of window_handle out of WINDOWSmain.h since it's no longer included by Platforms/Win32/include.h as a result of the fact that it includes windows.h and causes identifier conflicts. It is kept only to hold implementation state for the Win32 platform.

**Comment by RobertBColton** [comment]
Also, there's very little history on the external functions, try taking a look at it and see if you can get something to work if it's nothing to do with the above.
https://github.com/enigma-dev/enigma-dev/commits/master/ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSexternals.cpp

**Comment by time-killer-games** [comment]
@RobertBColton on it as we speak!

**Comment by time-killer-games** [comment]
@RobertBColton Here's the error that shows in the console btw 

`ERROR in some action of some event for object 0, instance id 100001: Attempting to access pointer variable as string`

**Comment by time-killer-games** [comment]
@RobertBColton I just tried all previous versions and they all have the same problem, minus the above error.

**Comment by RobertBColton** [comment]
That might be bad news for us, so if none of the above turns anything up, it could have to do with me adding pointer overloading for the var type. The var constructor may be confusing `const char*` and `void*` meaning it thinks all strings are actually pointers. Perhaps review f720a836e2e252be6764a12608a2a095c4721d23 and or consult with @JoshDreamland because this is the last of my ideas on this subject.

**Comment by time-killer-games** [comment]
@RobertBColton I would also like to point out that numbers are sometimes showing the wrong values as well. This I initially didn't entertain the possibility of, because I'm not used to it happening. But sure enough in some random cases completely wrong values for numbers are passed to arguments or returned by functions when printed or displayed somehow visually.

**Comment by RobertBColton** [comment]
Are you still talking about DLLs only or do you now mean games in general, e.g scripts?

**Comment by RobertBColton** [comment]
If you mean games in general, you might be getting affected by the #1448 GameData merge. If you'll notice, the Travis CI is also failing right now from it and we're trying to figure out why. There's also regressions in the compiler plus random epsilon errors.

**Comment by time-killer-games** [comment]
@RobertBColton I'm just talking about calling DLL's

**Comment by RobertBColton** [comment]
Ok, it could still be related to GameData anyway. Perhaps also review the history for DlgMod.
https://github.com/enigma-dev/enigma-dev/commits/master/ENIGMAsystem/SHELL/Widget_Systems/DlgMod

I'm out of ideas for now otherwise. The idea is to find where it was working and where it was that it stopped working. If you had enigma-dev locally I would tell you to be on the latest master branch and do `git reset --hard HEAD~1` (go back and test 1 commit at a time) until you are on a commit where DlgMod starts working again.

**Comment by time-killer-games** [comment]
@RobertBColton the history of DlgMod does not effect the DLL side. The DLL hasn't changed at all other than I added 3 functions since the time it was working. The test project does not use the DlgMod Widget System but rather uses scripts containing the external_* family of EDL functions to call the DLL within the project itself as aposed to using any Widget System whatsoever. The reason DlgMod doesn't work I am fully convinced is because of the same reason using the DLL in your EDL doesn't work; the external_* functions are broken somehow, I just wish we knew how or what the problem was and how to fix it.

**Comment by RobertBColton** [comment]
Clone a copy of ENIGMA and do the git reset I mentioned until you find the commit where DlgMod works again, that's the way to find out.

**Comment by time-killer-games** [comment]
Ok, will do! :)

**Comment by time-killer-games** [comment]
@RobertBColton I've gone as far back as when sorlok fixed OS X and everything I tried since then has either produced the same issue or emake segfaults during build. I give up.

**Comment by RobertBColton** [comment]
You were probably missing to `make clean` and `make all -j4` on some of the commits, but sure. Just leave the issue, somebody can look when they have time.

**Comment by RobertBColton** [comment]
Alright, so I finally downloaded the original project you provided. I didn't before because I didn't realize you made it that easy to just run it. Anyway, I wasn't actually able to reproduce any issues. I caught a few things though, one being the file filter on open dialogs seems to be corrupt.

![Open File Dialog Filter Corrupted](https://user-images.githubusercontent.com/3212801/53016816-7e0caf80-341c-11e9-90c5-6cc8b6d9cc74.png)

**Comment by RobertBColton** [comment]
Now, after a little more investigating, I did find something that is in fact ENIGMA's fault. The debug error messages are actually correct, but let me explain. They are coming from the var4 source file.
https://github.com/enigma-dev/enigma-dev/blob/ec519991d36435e01ed11a55c7367a9605b96451/ENIGMAsystem/SHELL/Universal_System/var4.cpp#L32

The specific thing that triggers them in your example is this:
```cpp
external_call(external_define("DialogModule.dll", "window_get_caption_legacy", dll_cdecl, ty_string, 1, ty_real), window_handle());
```

Basically, because the compatibility setting is broke, and even if it wasn't `window_handle()` returns a var pointer in GMS compliance mode. You'll recall that I made specifically this change for you. But when you define your external call, you are saying that the type of the `window_handle()` is real, when it's not. When I added variant pointer overloading to ENIGMA, I added the type `ty_pointer` but because this may be incompatible with GMS, I want to look at some way of fixing this for extensions.

**Comment by RobertBColton** [comment]
@time-killer-games So that leaves me with one question, what happens in GMSv1.4 if you pass `window_handle()` to a dll as `ty_real`? If that works, then ENIGMA needs to change something, if it doesn't then we should close this and just fix #1461 and the compatibility setting.

**Comment by time-killer-games** [comment]
@RobertBColton you just pointed out the issues I was talking about. Also notice the second screenshot you provided it is also using the title argument for the fname argument, this is also wrong.

**Comment by time-killer-games** [comment]
@RobertBColton calling window_get_caption_legacy is a ty_real because it is meant for the GM678 compatibility setting. if you change window_get_caption_legacy to window_get_caption and ty_real to ty_string in the scripts that is meant for the GMStudio compatibility.

**Comment by time-killer-games** [comment]
@RobertBColton You point out things that are wrong with the DLL calling and then you say everything on Enigma's end is fine. But that's not true. Those problems don't exist anywhere but in Enigma from my testing. And I've tested it in a lot of languages and software.

**Comment by RobertBColton** [comment]
Actually, you have that backwards, but sure, I did not know that it was specifically the filter the was wrong in ENIGMA. I was just looking at the fact that 99% of the rest of it was working fine, so I was a little confused. Anyway, the window handle is just one part of the cast problem. I don't think that explains the file filter, so I wonder what is causing that.

**Comment by time-killer-games** [comment]
The issue as I already stated is also present in the _ext() functions' fname argument. The fact any of the strings are working appears to be a fluke if you ask me. For instance, some of the input dialogs have a corrupt str argument for me. I'm sure if Josh tested This, he would also get a different result than the two of us.

**Comment by time-killer-games** [comment]
@RobertBColton I updated the dll in the sample project to the latest version. It won't make a difference, however.

**Comment by time-killer-games** [comment]
@RobertBColton I updated the dll in the sample project to the latest version.

**Comment by RobertBColton** [comment]
Ok, after speaking with Rusky about this and HitCoder, I'm pretty much of the opinion that this is caused by the fact that you've compiled the dll with MSVC. While this is fine for GMS use, because it uses MSVC, there is likely ABI compatibility differences between MinGW and MSVC that are causing these issues. I mean, this was my guess because I also know that when you build a DLL plugin for Qt Creator, just as an example, you must use the same compiler for the DLL that the Qt Creator version you've installed was compiled with. Simple way to confirm is to build the DLL in MinGW and see if this problem goes away.

**Comment by RobertBColton** [comment]
Closing as resolved by #1990.
